packages/react/src/extend-theme.ts

```diff
@@ -1,13 +1,33 @@
 import defaultTheme, { Theme } from "@chakra-ui/theme"
 import { isFunction, mergeWith } from "@chakra-ui/utils"
+import { ColorHues } from "@chakra-ui/theme/dist/types/foundations/colors"
+
+type ThemeExtensionTypeHints = {
+  colors: Record<string, Partial<ColorHues> | string> // typehints for color definitions
+}
+
+/**
+ * Represents a loose but specific type for the theme override.
+ * It provides autocomplete hints for extending the theme, but leaves room
+ * for adding properties.
+ */
+type DeepThemeExtension<ThemeObject, TypeHints> = {
+  [Key in keyof ThemeObject]?:
+    | Omit<DeepThemeExtension<ThemeObject[Key], TypeHints>, keyof TypeHints> // recursive type clone
+    | (ThemeObject[Key] extends (...args: any[]) => any
+        ? Partial<ReturnType<ThemeObject[Key]>>
+        : ThemeObject[Key]) // allow function or object
```

Comment:
Would we also want this to be `Partial`?

Reply:
I think is covered by the "escape hatch" `Record<string, any>` one line below, but I will add `Partial` for completeness!👍